core/src/main/scala/org/apache/spark/deploy/worker/StandaloneWorkerShuffleService.scala

```diff
@@ -35,7 +35,10 @@ private[worker]
 class StandaloneWorkerShuffleService(sparkConf: SparkConf, securityManager: SecurityManager)
   extends Logging {
 
-  private val enabled = sparkConf.getBoolean("spark.shuffle.service.enabled", false)
+  // Check both if shuffle service is enabled, and that the worker should actually host the
+  // shuffle service in that case. (The latter is currently only used for testing.)
+  private val enabled = sparkConf.getBoolean("spark.shuffle.service.enabled", false) &&
+    sparkConf.getBoolean("spark.worker.shouldHostShuffleServiceIfEnabled", true)
```

Comment:
FYI this change will conflict with https://github.com/apache/spark/pull/4990, which makes this a general class that Mesos can also use. If you follow my suggestion of just disabling this for local cluster then this doesn't need to change.